File: surface/widgets/gtk/components/widgets/primitives/ScrollLock.ts
// ============================================================
// Clef Surface GTK Widget — ScrollLock
//
// Scroll interception wrapper that prevents scrolling of the
// content beneath when active. Uses Gtk.ScrolledWindow with
// policy controls to lock/unlock scrolling.
//
// Adapts the scroll-lock.widget spec: anatomy (root), states
// (locked, unlocked), and connect attributes to GTK4 rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Props ---------------

export interface ScrollLockProps {
  locked?: boolean;
  child?: Gtk.Widget | null;
}

// --------------- Component ---------------

/**
 * Creates a GTK4 ScrolledWindow that locks/unlocks scrolling
 * based on the locked state.
 */
export function createScrollLock(props: ScrollLockProps = {}): Gtk.Widget {
  const { locked = false, child = null } = props;

  const scrolled = new Gtk.ScrolledWindow();

  if (locked) {
    scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.NEVER);
  } else {
    scrolled.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.AUTOMATIC);
  }

  if (child) {
    scrolled.set_child(child);
  }

  return scrolled;
}

File: surface/widgets/winui/components/widgets/primitives/ScrollLock.cs
// ============================================================
// Clef Surface WinUI Widget — ScrollLock
//
// Prevents parent-level scrolling when active. In WinUI 3,
// scrolling is controlled by ScrollViewer. This component
// wraps content and manipulates the parent ScrollViewer mode
// when locked. When inactive, children render without override.
//
// Adapts the scroll-lock.widget spec: anatomy (root), states
// (unlocked, locked), and connect attributes (data-part,
// data-state, data-scroll-lock) to WinUI 3 rendering.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace Clef.Surface.WinUI.Widgets.Primitives;

public sealed class ClefScrollLock : UserControl
{
    public static readonly DependencyProperty IsActiveProperty =
        DependencyProperty.Register(nameof(IsActive), typeof(bool), typeof(ClefScrollLock),
            new PropertyMetadata(false, OnPropertyChanged));

    public bool IsActive { get => (bool)GetValue(IsActiveProperty); set => SetValue(IsActiveProperty, value); }

    private readonly ScrollViewer _scrollViewer;
    private readonly ContentPresenter _presenter;

    public ClefScrollLock()
    {
        _presenter = new ContentPresenter();
        _scrollViewer = new ScrollViewer
        {
            Content = _presenter,
            HorizontalScrollBarVisibility = ScrollBarVisibility.Disabled,
            VerticalScrollBarVisibility = ScrollBarVisibility.Auto
        };
        Content = _scrollViewer;
        UpdateVisual();
    }

    public void SetLockContent(UIElement element)
    {
        _presenter.Content = element;
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        => ((ClefScrollLock)d).UpdateVisual();

    private void UpdateVisual()
    {
        _scrollViewer.VerticalScrollMode = IsActive ? ScrollMode.Disabled : ScrollMode.Auto;
        _scrollViewer.HorizontalScrollMode = IsActive ? ScrollMode.Disabled : ScrollMode.Auto;
    }
}
